File: 019-Remove-Nth-Node-From-End-of-List/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        fastPtr  = head
        slowPtr = head
        prevPtr  = None
        while n > 0:
            fastPtr = fastPtr.next
            n -= 1
        while fastPtr !=  None:
            fastPtr, prevPtr, slowPtr = fastPtr.next, slowPtr, slowPtr.next
        if slowPtr == head:
            head = slowPtr.next
        else:
            prevPtr.next = slowPtr.next
        del slowPtr
        return head

File: 019-Remove-Nth-Node-From-End-of-List/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode * scanPtr = head;
        ListNode * nthPtr = head;
        ListNode * prevPtr = NULL;
        int i = 0;
        // Because of that given n will always be valid, i do not do any kind of checking
        // Use two pointers, the fast pointer go n steps in advance, then the fast pointer and the slow pointer
        // go together, by doing this, when fast pointer run into the end of list, slow pointer will at the n-th node
        while (i < n){
            scanPtr = scanPtr->next;
            i++;
        }
        while(scanPtr != NULL){
            scanPtr = scanPtr->next;
            prevPtr = nthPtr;
            nthPtr  = nthPtr->next;
        }
        if(nthPtr == head)
            head = nthPtr->next;
        else
            prevPtr->next = nthPtr->next;
        delete nthPtr;
        return head;
    }
};
